### PR TITLE
Decouple encoding from query building

### PR DESCRIPTION
**Pull request by pnwpedro**
## Problem

In the throws of rapid development, query building and encoding were coupled.

## Solution

Decouple them by embedding the encoding protocol for a query builder in the encoding module rather than letting the builder encode itself and take a dependency on the encoder.

## Result

What will change as a result of your pull request? Note that sometimes this section is unnecessary because it is self-explanatory based on the solution.

## Testing

* Updated unit tests

### CODE REVIEW

File: fauna/query_builder.py [hunk 1]
```diff
@@ -2,15 +2,14 @@
...
 
 
 class QueryBuilder(abc.ABC):
     """An abstract class for query builders that build to the FQL Query Template wire protocol.
     """
 
     @abc.abstractmethod
-    def to_query(self) -> Mapping[str, Sequence[Any]]:
+    def fragments(self) -> Sequence['Fragment']:
         """An abstract method for converting a builder to query template wire protocol.
```
Comment (pnwpedro) [inline fauna/query_builder.py:13]: Oops, I forgot some docstring updates

File: fauna/query_builder.py [hunk 2]
```diff
@@ -63,39 +57,18 @@ class LiteralFragment(Fragment):
...
+        :returns: The stored value.
         """
         return self._val
 
 
-class QueryFragment(Fragment):
-    """A concrete :class:`Fragment` representing a subquery within a template. For example, if a template contains a
-    variable ```${foo_query}```, and a :class:`QueryBuilder` is provided for foo, then it should be wrapped as a
-    :class:`QueryFragment`.
+class QueryInterpolationBuilder(QueryBuilder):
```
Comment (ptpaterson) [inline fauna/query_builder.py:68]: Do we even need "builders" now? Can this still be a `QueryFragment` whose `_val` is a `List[Fragment]`?

Other builder method have been taken away. What's left is `fragments` which a simple getter. At this point the `QueryBuilder` class looks a lot like a `Fragment` sub class.

File: fauna/query_builder.py [hunk 3]
```diff
@@ -104,32 +77,15 @@ def __init__(self, fragments: Optional[List[Fragment]] = None):
         self._rendered = None
```
Comment (ptpaterson) [inline fauna/query_builder.py:77]: Does `fragments` really need to be Optional in the `__init__` method? If it's required, then this `QueryBuilder` subclass would look even more like a `Fragment`.

Comment (pnwpedro) [reply fauna/query_builder.py:77]: I removed append for now, but I think it's fine to keep this optional. It gives us greater flexibility with how we use these builders. e.g. we can add append back and init and append afterward, which maybe we (or someone) will want depending on how they're building their queries.

File: fauna/query_builder.py (hunk 2 above)
Comment (pnwpedro) [reply fauna/query_builder.py:68]: The builder is the contract the for the query API, which is different from a fragment.

For instance, if we move this to be just a Fragment, then the API either needs to accept a QueryFragment or a Fragment. The latter may not always make sense (i.e. a Fragment isn't by definition always a valid query. The former, for it to be extensible, would have to be abstract and then we'd implement QueryInteroplationFragment.

I think keeping these ideas separate means we can have an extensible yet safe API contract and we don't run the risk of Fragments and their uses being in conflict with a query builder and their uses.

Comment (pnwpedro) [reply fauna/query_builder.py:68]: We could rename these if you think that'd be an improvement, though. Like `QueryInteropolation` and an abstract `Query`

Comment (pnwpedro) [reply fauna/query_builder.py:68]: Taking this a bit further, maybe removing the abstract class altogether for now and treat QueryInterpolation as a collection of fragments. Then another query type doesn't necessarily need to rely on fragments.

**Review by ptpaterson — APPROVED** [review]

**Review by cynicaljoy — APPROVED** [review]

**Pull request merged**